src/common/expression/src/evaluator.rs

```diff
@@ -494,33 +513,51 @@ impl<'a> ConstantFolder<'a> {
                 generics,
                 args,
             } => {
-                let (mut args_expr, mut args_domain) = (Vec::new(), Vec::new());
-
+                let (mut args_expr, mut args_domain) = (Vec::new(), Some(Vec::new()));
                 for arg in args {
                     let (expr, domain) = self.fold(arg);
-                    if let Some(domain) = domain {
-                        args_expr.push(expr);
-                        args_domain.push(domain);
-                    } else {
-                        return (expr, None);
-                    }
+                    args_expr.push(expr);
+                    args_domain = args_domain.zip(domain).map(|(mut domains, domain)| {
+                        domains.push(domain);
+                        domains
+                    });
                 }
 
-                let func_domain = (function.calc_domain)(&args_domain, generics);
-                let func_expr = func_domain
-                    .as_ref()
-                    .and_then(Domain::as_singleton)
-                    .map(|scalar| Expr::Constant {
-                        span: span.clone(),
-                        scalar,
-                    })
-                    .unwrap_or_else(|| Expr::FunctionCall {
-                        span: span.clone(),
-                        id: id.clone(),
-                        function: function.clone(),
-                        generics: generics.clone(),
-                        args: args_expr,
-                    });
+                let func_domain =
+                    args_domain.and_then(|domains| (function.calc_domain)(&domains, generics));
+                let all_args_is_scalar = args_expr.iter().all(|arg| arg.as_constant().is_some());
+
+                let func_expr = Expr::FunctionCall {
+                    span: span.clone(),
+                    id: id.clone(),
+                    function: function.clone(),
+                    generics: generics.clone(),
+                    args: args_expr,
+                };
+
+                if let Some(scalar) = func_domain.as_ref().and_then(Domain::as_singleton) {
+                    return (
+                        Expr::Constant {
+                            span: span.clone(),
+                            scalar,
+                        },
+                        func_domain,
+                    );
+                }
+
+                if all_args_is_scalar {
+                    let chunk = Chunk::empty();
+                    let evaluator = Evaluator::new(&chunk, self.context.clone());
+                    if let Ok(Value::Scalar(scalar)) = evaluator.run(&func_expr) {
```

Comment:
`Value::Column(c) if c.len() == 1` should be cared.

There will be functions that accept more args than 3 like `elt`.